agent/src/main/java/org/openremote/agent/protocol/dmx/artnet/ArtnetClientProtocol.java

```diff
@@ -0,0 +1,502 @@
+                    if(lightId != -1 && universe != -1) {
+                        if(artnetLightMemory.stream().anyMatch(light -> light.getLightId() == lightId)) {
+                            artnetLightMemory.stream().filter(light -> light.getLightId() == lightId).findFirst().ifPresent(artnetLight -> artnetLightMemory.remove(artnetLight));
+                        }
+                    }
+                }
+            }
+        }
+        synchronized (protocolMessageConsumers) {
+            protocolMessageConsumers.compute(protocolConfiguration.getReferenceOrThrow(), (ref, consumers) -> {
+                if (consumers != null) {
+                    consumers.removeIf((attrRefConsumer) -> attrRefConsumer.key.equals(attribute.getReferenceOrThrow()));
+                }
+                return consumers;
+            });
+        }
+
+    }
+
+    @Override
+    protected void onMessageReceived(AttributeRef protocolRef, ArtnetPacket packet) {
+        List<Pair<AttributeRef, Consumer<ArtnetPacket>>> consumers;
+        synchronized (protocolMessageConsumers) {
+            consumers = protocolMessageConsumers.get(protocolRef);
+            if (consumers != null) {
+                consumers.forEach(c -> {
+                    if (c.value != null) {
+                        c.value.accept(packet);
+                    }
+                });
+            }
+        }
+    }
+
+    @Override
+    protected ArtnetPacket createWriteMessage(AssetAttribute protocolConfiguration, AssetAttribute attribute, AttributeEvent event, Value processedValue) {
+        //Todo check for group later here
+        AttributeRef attributeRef = attribute.getReference().orElse(null);
+        if(attributeRef != null) {
+            AssetAttribute linkedAttribute = getLinkedAttribute(attributeRef);
+            if(linkedAttribute != null) {
+                String parentAssetId = linkedAttribute.getAssetId().orElse(null);
+                if(parentAssetId != null) {
+                    Asset parentAsset = assetService.findAsset(parentAssetId);
+                    Attribute universeAttribute = parentAsset.getAttribute("Universe").orElse(null);
+                    Attribute lightAttribute = parentAsset.getAttribute("Id").orElse(null);
+                    if(universeAttribute != null && lightAttribute != null) {
+                        int universeId = universeAttribute.getValueAsInteger().orElse(-1);
+                        int lightId = lightAttribute.getValueAsInteger().orElse(-1);
+                        if(universeId != -1 && lightId != -1) {
+                            ArtnetLight updatedLight = (ArtnetLight) artnetLightMemory.stream().filter(light -> light.getLightId() == lightId).findFirst().orElse(null);
+                            if(updatedLight != null) {
+                                ArtnetLightState oldLightState = (ArtnetLightState) updatedLight.getLightState();
+                                ObjectMapper mapper = new ObjectMapper();
```

Comment:
Use `Container.JSON`

Reply:
There are also helper methods in `Values` `convertFromValue` and `convertToValue`

Reply:
> There are also helper methods in `Values` `convertFromValue` and `convertToValue`

We have chosen to check each individual field in the json-object to make sure whether it is present or not. This is important as we do not want to update the lightstate if the provided json is invalid (is missing a property).

Reply:
Use convertFromValue to get the lightstate object then check that object for validity